Link shipping form fields to shipping rates to load them.

senadir opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR is a continuation from yesterday's mega PR, it fixes the problem with shipping rates not updating when you change a value, it now works fine, and is debounced, there are some updates to how `useShippingRates` work now, it is required that you pass the shape of your fields for them to be watched and updated accordingly. 
<!-- Reference any related issues or PRs here -->
Fixes #1883

### How to test the changes in this Pull Request:

1. go to a page/post with Cart or Checkout block, and a sit with shipping rates.
2. enter your shipping informations.
3. watch that changing first name, last name and street address does not trigger a rate update.
4. changing country, state, city or postcode will trigger an update.
5. updates are debounced.

senadir commented on assets/js/base/hooks/use-shipping-rates.js:
```diff
@@ -1,40 +1,63 @@
-			updateShippingAddress,
-		};
-	}, [] );
+	const derivedAddress = shippingRates[ 0 ]?.destination;
```
we prefill `derivedAddress` with the address we got from updated shipping rates.

senadir commented on assets/js/base/hooks/use-shipping-rates.js:
```diff
@@ -1,40 +1,63 @@
-		};
-	}, [] );
+	const derivedAddress = shippingRates[ 0 ]?.destination;
+	const initialAddress = { ...addressFields, ...derivedAddress };
```
we create a new initial state from fields shape + existing fields

senadir commented on assets/js/base/hooks/use-shipping-rates.js:
```diff
@@ -1,40 +1,63 @@
-	}, [] );
+	const derivedAddress = shippingRates[ 0 ]?.destination;
+	const initialAddress = { ...addressFields, ...derivedAddress };
+	const shippingAddressReducer = ( state, address ) => ( {
```
a reducer takes new addresses and merge them with old ones

senadir commented on assets/js/base/hooks/use-shipping-rates.js:
```diff
@@ -1,40 +1,63 @@
+		shippingAddressReducer,
+		initialAddress
+	);
+	const [ debouncedShippingAddress ] = useDebounce( shippingAddress, 400 );
```
we debounce the reducer value.

senadir commented on assets/js/base/hooks/use-shipping-rates.js:
```diff
@@ -1,40 +1,63 @@
+			debouncedShippingAddress.country
+		) {
+			updateShippingAddress( debouncedShippingAddress );
+		}
```
we only trigger a rate update if a relevant value has changed (country, state, city, postcode) and if a country is set.

senadir commented on assets/js/base/utils/address.js:
```diff
@@ -0,0 +1,14 @@
+	state,
+	city,
+	postcode,
+} );
```
plunk takes an address object and return needed fields, this is required for comparison with shallow equal and to track needed fields, we might supply this to the hook if we want to (not blocking)

senadir commented on assets/js/blocks/cart-checkout/cart/full-cart/index.js:
```diff
@@ -103,8 +103,17 @@ const Cart = ( {
+		shippingAddress,
+		setShippingAddress,
+		shippingRatesLoading,
+	} = useShippingRates( defaultAddress );
```
we refactor Cart block to use the new muse.

senadir commented on assets/js/blocks/cart-checkout/checkout/block.js:
```diff
@@ -72,6 +70,16 @@ const Block = ( { attributes, isEditor = false, shippingRates = [] } ) => {
+	} = useShippingRates(
+		Object.fromEntries(
+			Object.entries( addressFields ).map( ( [ key ] ) => [ key, '' ] )
+		)
```
the current shape of `addressFields` is key => object of meta things, we only something like
```js
{
	country: '',
	state: '',
	city: '',
	postcode: '',
	};
```
so we remap the object to this.

senadir replied:
the goal is to always and watch and send the fields being used in the form, this ensures that if someone adds custom fields in the future, it should also work, in the cart, we don’t have any sort of dynamic fields yet (the fields are hardcoded in https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/4ddcf447be4f9bce7e973a7c91d9da148d1d2f3d/assets/js/base/components/shipping-calculator/address.js#L22

we might export them somewhere and use them here.

there is also the logic of turning an array to an object, this I will move to the hook.

senadir replied:
fixed in a73d39d14

senadir commented on assets/js/base/hooks/use-shipping-rates.js:
```diff
@@ -1,40 +1,63 @@
+		...state,
+		...address,
+	} );
+	const [ shippingAddress, setShippingAddress ] = useReducer(
```
One issue that is not blocking but didn’t understand yet is that `useReducer` initial value doesn’t work like `useState`, it seems `shippingAddress` is returning empty values, even if `initialAddress` has values in it, this means on page load, no address will be in the checkout shipping fields, this is not blocking but we should fix it.

Aljullu commented on assets/js/base/hooks/use-shipping-rates.js:
```diff
@@ -1,40 +1,64 @@
+		( select ) => select( storeKey ).areShippingRatesLoading(),
+		[]
+	);
+	const { updateShippingAddress } = useDispatch( storeKey );
```
Before, this was inside the `useSelect()`. Is there any benefit on having it outside? Not saying I prefer one way or the other, just asking to help me understand. :slightly_smiling_face:

senadir replied:
no benefit from having it inside or outside, this simply avoids us returning it from useSelect `results` then taking it off again, according to Darren there should be no difference 000&cid=C8X6Q7XQU-slack-C8X6Q7XQU

Aljullu commented on assets/js/base/hooks/test/use-shipping-rates.js:
```diff
@@ -19,11 +19,13 @@ describe( 'useShippingRates', () => {
 	let registry, mocks, renderer;
 	const getProps = ( testRenderer ) => {
 		const {
-			shippingRates,
```
Couldn't we just leave the `shippingRates` in the tests? Or you removed them for some reason?

senadir replied:
well at first I did because I also removed shippingRates from useShipping for some reason? and returned it. refactored

senadir commented:
@Aljullu so I addressed all of your issues except the one in which shipping rates are not returning, this seems to be an issue from the REST API (no rates are being returned for countries that have states/cities defined in Woo) so I tested with Spain, it has cities, it won’t work until you define a city, I tested with US, it didn’t work, I tested with Algeria, Morocco and France, non of which have states in Woo core, and they all worked fine, for other countries, it simply doesn’t return, worth opening an issue and seeing why.
![image](https://user-images.githubusercontent.com/6165348/76161544-b15bd280-6134-11ea-800c-8795b12ef5f9.png)

Aljullu commented:
> @Aljullu so I addressed all of your issues except the one in which shipping rates are not returning, this seems to be an issue from the REST API (no rates are being returned for countries that have states/cities defined in Woo) so I tested with Spain, it has cities, it won’t work until you define a city, I tested with US, it didn’t work, I tested with Algeria, Morocco and France, non of which have states in Woo core, and they all worked fine, for other countries, it simply doesn’t return, worth opening an issue and seeing why.

Thanks for finding steps to reproduce and opening an issue @senadir.

senadir commented:
I followed up on the initial issue and fixed it in our repo while it’s fixed in Woo Core.
